dd-sdk-android/src/main/kotlin/com/datadog/android/rum/internal/monitor/DatadogRumMonitor.kt

```diff
@@ -215,6 +243,73 @@ internal class DatadogRumMonitor(
 
     // region Internal
 
+    private fun updateAttributesWithActionId(
+        attributes: Map<String, Any?>
+    ): Map<String, Any?> {
+        val actionId = getActiveActionId()
+        return if (actionId == null) {
+            attributes
+        } else {
+            attributes.toMutableMap()
+                .apply {
+                    put(RumEventSerializer.TAG_EVENT_USER_ACTION_ID, actionId.toString())
+                }
+        }
+    }
+
+    private fun getActiveActionId(): UUID? {
+        val actionEvent = activeActionEvent
+        val actionData = activeActionData
+        val isWithinActionScope = isWithinActionScope()
+
+        return if (
+            actionEvent != null &&
+            actionData != null &&
+            isWithinActionScope
+        ) {
+            val nanoTime = System.nanoTime()
+            lastActionRelatedEventNs.set(nanoTime)
+            actionData.id
+        } else {
+            sendUnclosedAction()
+            null
+        }
+    }
+
+    private fun isWithinActionScope(): Boolean {
+        val nanoTime = System.nanoTime()
+        val lastAction = lastActionRelatedEventNs.get()
+        val isLastEventRecent = (nanoTime - lastAction) < ACTION_INACTIVITY_NS
+        ongoingUserActionResources.removeAll { it.get() == null }
+        val hasUnclosedResources = ongoingUserActionResources.isNotEmpty()
+        return isLastEventRecent || hasUnclosedResources
+    }
+
+    private fun sendUnclosedAction() {
+        val lastAction = lastActionRelatedEventNs.get()
+        val actionEvent = activeActionEvent
+        val actionData = activeActionData
+        sendUserAction(actionEvent, actionData, lastAction)
+        activeActionEvent = null
+        activeActionData = null
+    }
+
+    private fun updateUserActionResources(key: Any) {
+        val nanoTime = System.nanoTime()
+        val actionEvent = activeActionEvent
+        val actionData = activeActionData
+
+        if (actionEvent != null && actionData != null) {
```

Comment:
you don't want to remove the resource no matter what as it was closed ?

Reply:
true